[Merged by Bors] - feat(linear_algebra): Smith normal form for submodules over a PID

Vierkantor opened the pull request:
This PR expands on `submodule.basis_of_pid` by showing that this basis can be chosen in "Smith normal form". That is: if `M` is a free module over a PID `R` and `N ≤ M`, then we can choose a basis `bN` for `N` and `bM` for `M`, such that the inclusion map from `N` to `M` expressed in these bases is a diagonal matrix in Smith normal form.

The rather gnarly induction in the original `submodule.basis_of_pid` has been turned into an even gnarlier auxiliary lemma that does the inductive step (with the induction hypothesis broken into pieces so we can apply it part by part), followed by a re-proven `submodule.basis_of_pid` that picks out part of this inductive step. Then we feed the full induction hypothesis, along with `submodule.basis_of_pid` into the same induction again, to get `submodule.exists_smith_normal_form_of_le`, and from that we conclude our new results `submodule.exists_smith_normal_form` and `ideal.exists_smith_normal_form`.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

 - [x] depends on: #8567 

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
feat(linear_algebra): Smith normal form for submodules over a PID

This PR expands on `submodule.basis_of_pid` by showing that this basis can be
chosen in "Smith normal form". That is: if `M` has a basis and `N ≤ M`, then we
can choose a basis `bN` for `N` and `bM` for `M`, such that the inclusion map
from `N` to `M` expressed in these bases is a diagonal matrix in Smith normal
form.

jcommelin commented on src/linear_algebra/free_module_pid.lean:
```diff
@@ -79,12 +88,33 @@ begin
 
+lemma eq_bot_of_generator_maximal_range_eq_zero {N O : submodule R M} (b : basis ι R O)
+  (hNO : N ≤ O)
+  {ϕ : O →ₗ[R] R} (hϕ : ∀ (ψ : O →ₗ[R] R), (ϕ.comp $ of_le hNO).range ≤ (ψ.comp $ of_le hNO).range →
```
It seems that you are using this assumption `hϕ` several times, and it looks quite intimidating. Do you think it can be packaged behind a nice and readable name?

Vierkantor replied:
Informally stated, the condition on `φ` is that `φ(N)` is maximal among all such images `ψ(N)`. I can't find a definition for "is a maximal element", for example Zorn's lemma is stated in the same `x ≤ y → x = y` way. But for `(φ.comp (of_le hNO)).range` (`φ(N)`) I can definitely make up a name like `φ.submodule_image hNO`.

jcommelin commented on src/linear_algebra/free_module_pid.lean:
```diff
@@ -404,6 +608,109 @@ noncomputable def module.free_of_finite_type_torsion_free' [module.finite R M]
 module.free_of_finite_type_torsion_free module.finite.exists_fin.some_spec.some_spec
 
+/-- If `M` is finite free over a PID `R`, then any submodule `N` is free
+and we can find a basis for `M` and `N` such that the inclusion map is a diagonal matrix.
```
It's not just a diagonal matrix, it's in SNF. I think the docstring should also mention that.

jcommelin commented on src/linear_algebra/free_module_pid.lean:
```diff
@@ -404,6 +608,109 @@ noncomputable def module.free_of_finite_type_torsion_free' [module.finite R M]
+-/
+theorem submodule.exists_smith_normal_form_of_le [fintype ι]
+  (b : basis ι R M) (N O : submodule R M) (N_le_O : N ≤ O) :
+  ∃ (n o : ℕ) (hno : n ≤ o) (bO : basis (fin o) R O) (bN : basis (fin n) R N) (a : fin n → R),
```
Does it make sense to define a `SNF` predicate, and then use that as the conclusion of this theorem?

Vierkantor replied:
Thinking about how I'm going to apply it, I'll even turn it into a structure.

Vierkantor replied:
I defined a `basis.smith_normal_form` structure for a pair of bases, for the whole module and a submodule, that are in SNF.

jcommelin reviewed (commented):
gnarly indeed!!!

Vierkantor committed:
Update docstring

Vierkantor committed:
@[simps] this def

Vierkantor committed:
Make a `basis.smith_normal_form` structure

Instead of returning a big `∃`, now we can return a structure,
in the noncomputable defs `{submodule,ideal}.smith_normal_form[_of_le]`.
The `exists_smith_normal_form` theorems return the `∃` statements,
which might be more convenient to apply if you don't want to map the objects
into the correct submodule. (I couldn't find a way to make the structure
general enough that these maps aren't needed. Some sort of `scalar_tower`?)

Vierkantor committed:
Introduce an abbreviation for `φ(N)`, i.e. `φ.comp (of_le (_ : N ≤ _))`

Vierkantor committed:
Nicer names/comments

Vierkantor committed:
Reduce hypotheses

Vierkantor committed:
Fix docstrings referring to old name

Vierkantor committed:
Just nolint the inhabited instance, we need nontrivial assumptions to get one

jcommelin commented on src/linear_algebra/free_module_pid.lean:
```diff
@@ -404,6 +659,129 @@ noncomputable def module.free_of_finite_type_torsion_free' [module.finite R M]
+@[nolint has_inhabited_instance]
+structure basis.smith_normal_form (N : submodule R M) (ι : Type*) (n : ℕ) :=
+(bM : basis ι R M) (bN : basis (fin n) R N) (f : fin n ↪ ι)
+(a : fin n → R) (snf : ∀ i, (bN i : M) = a i • bM (f i))
```
You are doing something a bit more general right? Because you include the `n : nat` that allows you to do "partial(?)" SNF bases?

Unrelated: I think I would prefer having one structure field per line.

Vierkantor replied:
The point of using two different indices is that submodules that are not of full rank can still have SNF bases. That is, the matrix you end up is rectangular with `A i (f i) = a i` and `A i j = 0` otherwise. Or you can view it as a full `ι × ι` matrix with `A (f i) (f i) = a i` and `A i j = 0` otherwise, including on the diagonal outside of the image of `f`. And the reason for using `fin n` rather than an abstract type `κ` is that you can ensure `a i | a (i + 1)`.

Maybe "diagonal matrix" is a poor choice of words, I have changed it into "a (rectangular) matrix with `a` along the diagonal: in Smith normal form."

I have also reformatted the fields.

Vierkantor committed:
Don't need this assumption

Vierkantor committed:
Clarify docstring

Vierkantor committed:
Reformat fields

jcommelin reviewed (approved):
Thanks :tada:

bors d+
